design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,217 @@
+        resources:
+          requests:
+            storage: "1Gi"
+  network:
+    ports:
+    - name: http
+      port: 8080
+    - name: grpc
+      port: 26257
+  secure: false
+  cachePercent: 25
+  maxSQLMemoryPercent: 25
+  # A key/value list of annotations
+  annotations:
+  #  key: value
+```
+
+2. Create a portable `PostgreSQLInstanceClass`
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstanceClass
+metadata:
+  name: postgresql-standard
+  namespace: app-project1-dev
+classRef:
+  kind: ClusterClass
+  apiVersion: cockroachdb.rook.crossplane.io/v1alpha1
+  name: rook-cockroachdb
+  namespace: rook-infra-dev
+```
+
+3. Create a `PostgreSQLInstance` claim
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstance
+metadata:
+  name: app-postgresql
+  namespace: app-project1-dev
+spec:
+  classRef:
+    name: postgresql-standard
+  writeConnectionSecretToRef:
+    name: postgresqlconn
+  engineVersion: "9.6"
+```
+
+On creation of the `PostgreSQLInstance` claim, the claim controller would
+configure and create a `Cluster` (`cockroachdb.rook.crossplane.io/v1alpha1`)
+instance. Then, the managed controller would use the Crossplane `Cluster`
+instance to create a `Cluster` (`cockroachdb.rook.io/v1alpha1`) in the target
+Kubernetes cluster where Rook and the CockroachDB operator are installed. This
+workflow is similar to dynamic provisioning in a typical infrastructure
+provider:
+
+![K8s Native Providers](./images/k8s-native-providers.png)
+
+## Technical Design
+
+### Claim Reconciler
+
+The claim controllers could use the same shared [claim reconciler] that other
+existing infrastructure stacks use currently, providing their own
+`ManagedConfigurators` to configure the managed resource instance using the
+referenced class.
+
+### Scheduler Reconciler
+
+Because the scheduling behavior for most Kubernetes-native infrastructure
+components will look very similar, a shared scheduler reconciler should be added
+to [crossplane-runtime]. It will look almost identical to the reconciler used
+for `KubernetesApplication` [scheduler controller]. In fact, the
+`KubernetesApplication` scheduler controller should likely make use of the
+shared `crossplane-runtime` scheduler reconciler after it is implemented.
+
+### Managed Reconciler
+
+Though this design is motivated by the desire to support Rook, the pattern can
+be generalized to be applicable to other Kubernetes-native infrastructure
+providers. While most existing infrastructure provider stacks must create a
+client wrapper library to use the CRUD functions of the provider API, every
+Kubernetes-native infrastructure stack will use the [client-go] SDK for CRUD
+operations. Therefore, the pluggable `Observe`, `Create`, `Update`, `Delete`
+functions in the [crossplane-runtime] can likely have a common implementation
+for Kubernetes-native infrastructure stacks.
+
+This can be done via generic `NativeConnecter` and `nativeExternal` types in
```

Comment:
I think the connector will be almost identical for all of these k8s-native providers, but I agree that the external functions will likely be much harder to generalize. I wonder if we could add a `resource.WithNativeExternalConnector()` function and just pass it an `external` object that satisfies the external object but just automatically gets the k8s client injected into it from a shared connector implementation. All of these are sort of "nice-to-have" features, rather than imperative for functionality.

Reply:
I think I don't know how many k8s-native providers we can have currently or plan to have. That's why I'm worried that we add a generic functionality to crossplane-runtime that is used by one provider. If we were to use GCP's config connector and AWS service operator, I'd definitely go for it though.

Reply:
I think the primary scenario here is operators, not cloud provider declarative k8s APIs.  There are TONS of operators out there now and being able to integrate with the general scenario of the CRD API that is backed by operators is pretty compelling.  It completely opens up on-premises scenarios for being managed by Crossplane.

Likely, we'd end up with a generalized solution in the first pass here because Rook itself is 7+ different operators.  If we want to do S3 and Postgres types then that covers CockroachDB, Yugabyte, Ceph, EdgeFS, and Minio, or essentially 5 different operators.  Sounds like there could already be a case for generalizing this in the first pass :)

Reply:
> All of these are sort of "nice-to-have" features, rather than imperative for functionality.
Could we skip the "nice-to-have" features for now and circle back if the need arises?

Reply:
I believe the current doc is aligned on tackling Rook specifically to begin with, so I'm going to resolve this thread.